core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java

```diff
@@ -92,6 +94,15 @@ protected final void check(
     sql(sql).convertsTo(plan);
   }
 
+  // TODO(CALCITE-3571): Passes when RelBuilder#shouldMergeProject returns false.
+  @Ignore public void testJoinWithMergeProjectShouldParse() {
+    final String sql = "WITH query as (select empno, deptno as deptno11, deptno as deptno12 from emp)\n"
+        + "select query.deptno11, emp.deptno, query.deptno12\n"
+        + "FROM query\n"
+        + "JOIN emp ON (cast(query.empno as Integer) = cast(emp.empno as Integer))";
+    sql(sql).ok();
+  }
```

Comment:
IMHO, what you propose is only a workaround. Users might encounter this error if he wants to merge the project in RelBuilder. Instead of changing `RelBuilder#shouldMergeProject` to false, I am wondering if we can find the root cause and fix it.

Reply:
I totally agree. Still in the process of investigating this.
I also wonder if `Projects` will get merged anyways by `ProjectMergeRule` by the optimizer?